libsolidity/experimental/ast/TypeSystem.h

```diff
@@ -74,6 +74,12 @@ class TypeEnvironment
 	TypeSystem& typeSystem() { return m_typeSystem; }
 	TypeSystem const& typeSystem() const { return m_typeSystem; }
 
+	bool isFixedTypeVar(Type const& _typeVar) const;
+	bool isGenericTypeVar(Type const& _typeVar) const;
+	bool hasGenericTypeVars(Type const& _type) const;
```

Comment:
This should not be in ``TypeEnvironment``.
``TypeEnvironment`` and ``TypeSystem`` should have minimal interfaces - anything that can be built on top of those minimal interfaces auxiliarily should be done in ``TypeEnvironmentHelpers`` or ``TypeSystemHelpers``. (Yes, even if there may already be some sort of exception to that ;-))

Reply:
I moved `hasGenericTypeVars()` to `TypeEnvironmentHelpers`. The other two need to be here because there's no way to access `m_typeVariables` and `m_fixedTypeVariables` from the outside. Or should I just add getters for them?

Reply:
> `TypeEnvironment` and `TypeSystem` should have minimal interfaces

Why though? I have to say that the current way to access these helpers is pretty horrible. You have to create the object each time, which makes things verbose and inconvenient to use. And the only state they carry is really the original object so all their methods feel like they should just be members on that object.

Reply:
Well, ok, this finally puts it into a different perspective for me. I was thinking about it being a different kind of interface. I.e. only a loose, internal one for us, not such a strict boundary.

In any case, are `isGenericTypeVar()` and `isFixedTypeVar()` fine there OR should I expose the getter OR do you see some other way without providing this info outside the interface at all?

Reply:
They're fine - and they're better than a full getter for the underlying data structure even.

Reply:
One of them is enough, though - the other one is just its negation ;-). But yeah, that's maybe exaggerating strictness on this, it won't hurt too much to have both ;-).

Reply:
Well, not exactly negation. It also checks if it's really a type variable. But ok, easy enough to eliminate. I had only one place where I wanted `isGenericTypeVar()` so far so I replaced it.